### ISSUE TITLE
[HOLD][$1000] The main composer shouldn’t show when editing message

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15959]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open any chat
2. Edit message A and the keyboard is displayed
3. Observe that the main composer is disappeared
4. Continue clicking to edit message B
5. Cancel editting in the message A
6. Observe that the main composer appears

## Expected Result:
The main composer shouldn’t show when editing message

## Actual Result:
The main composer appears when editing message

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.84-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/225112956-6919167c-ca17-45b5-8acf-50e748322a50.mov

https://user-images.githubusercontent.com/43996225/225112988-fde33581-9e2d-4dba-a63d-482304a6d0f0.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @tienifr 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1678700304143029

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010062f578265fa155</li>
        <li>Upwork Job ID: 1638674490806386688</li>
        <li>Last Price Increase: 2023-04-06</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @stephanieelliott (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by kavimuru** [comment]
@stephanieelliott Sorry I have edited the actual result. Both are same videos. One is by the contributor (tienifr) and another is mine.

**Comment by stephanieelliott** [comment]
Perfect, thanks @kavimuru - makes way more sense now!

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~010062f578265fa155

**Comment by MelvinBot** [comment]
Current assignee @stephanieelliott is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @thienlnam (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The main composer shouldn’t show when user's in edit mode

### What is the root cause of that problem?

In https://github.com/Expensify/App/blob/8a04b2c9169a9294011535ce86b4c12018f7fa11/src/pages/home/report/ReportActionItemMessageEdit.js#L142-L155
We call `toggleReportActionComposeView` function whenever user click cancel button. 

### What changes do you think we should make in order to solve the problem?

On mWeb, onBlur function will fire if we click cancel button, but not on native platform.
That why I suggestion to have the new function called `shouldShowComposeInput`. It returns `isFocused` state if it's native platform, otherwise returns true.

Here is my suggestion:

```

deleteDraft() {
        this.debouncedSaveDraft.cancel();
        Report.saveReportActionDraft(this.props.reportID, this.props.action.reportActionID, '');
       if(shouldShowComposeInput(this.state.isFocused)){
        toggleReportActionComposeView(true, this.props.isSmallScreenWidth);
        ReportActionComposeFocusManager.focus();
        // Scroll to the last comment after editing to make sure the whole comment is clearly visible in the report.
        if (this.props.index === 0) {
            const keyboardDidHideListener = Keyboard.addListener('keyboardDidHide', () => {
                ReportScrollManager.scrollToIndex({animated: true, index: this.props.index}, false);
                keyboardDidHideListener.remove();
            });
        }
       }
    }

```

#### Result

https://user-images.githubusercontent.com/113963320/225114445-57f6ef74-ab42-4e14-bcc8-c288ad951855.mp4

**Comment by s77rt** [comment]
I think this is a _dangerous_ feature to implement. What if the user has a draft message somewhere that he can't see (an old message). After clicking Cancel the Main composer won't show since he is still in edit more on that other lost message.

@tienifr How would your proposal look on mWeb?

**Comment by thienlnam** [comment]
Yeah that's a good point - is there an actual problem with the main composer showing? If there is no problem with the keyboard I think we can do nothing here

**Comment by s77rt** [comment]
@tienifr There is a reason I asked your for a video on mWeb. I think there is a misunderstanding here. Looks you are trying to fix another issue.

**Comment by tienifr** [comment]
@s77rt The video below is what happens on mWeb. Pls let me know your though. Thanks 😁

https://user-images.githubusercontent.com/113963320/228174652-04c0347d-a9f1-4285-8969-276cdb86513d.mp4

**Comment by s77rt** [comment]
@tienifr The issue still exists on mWeb.

**Comment by tienifr** [comment]
@s77rt I think on mweb, after clicking the Cancel button, we always blur the input => we show the main composer again. We just need to hide the main composer when we're focusing on the edit composer right?

**Comment by s77rt** [comment]
@rushatgabhane @thienlnam I think we can keep this issue open. The title is a bit confusing. What we are looking to fix here is:
If you are editing message A and click cancel on message B. The main composer should not show and focus should be retained on message A that you are editing. IOW The main composer should't show clicking on Cancel while editing another message.

This is a safe feature to implement :grin:

**Comment by MelvinBot** [comment]
@rushatgabhane @thienlnam @stephanieelliott this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by ShogunFire** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When we are editing two comments and we stop editing one of them, the main composer is showing and it shouldn't



### What is the root cause of that problem?

In the method deleteDraft here

https://github.com/Expensify/App/blob/8a04b2c9169a9294011535ce86b4c12018f7fa11/src/pages/home/report/ReportActionItemMessageEdit.js#L142

We toggle the main composer and we give it focus.



### What changes do you think we should make in order to solve the problem?
 
 We can save the number of messages that are being edited (with a static field ?)
 
In deleteDraft we toggle the main composer and change the focus only when this is the only draft being edited.

### What alternative solutions did you explore? (Optional)

**Comment by stephanieelliott** [comment]
Not overdue, still discussing and evaluating proposals.

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1000

**Comment by thienlnam** [comment]
Updating price back down, we were still discussing whether this was an actual bug or not

**Comment by rushatgabhane** [comment]
> Not related to this issue but do we have a story behind that our App allow to have multiple editing messages?

I'd think you don't wanna lose your edit drafts

**Comment by MelvinBot** [comment]
@rushatgabhane, @thienlnam, @stephanieelliott Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by ShogunFire** [comment]
Hello @s77rt what do you think of my proposal ? I can make it more detailed if needed

**Comment by s77rt** [comment]
@ShogunFire I'm not the C+ here. @rushatgabhane will review your proposal shortly. But I don't think the use of the number of drafts is a good approach here. Your proposal will create the problem highlighted [here](https://github.com/Expensify/App/issues/15959#issuecomment-1483819425) (hidden comment).

If the current focused input is Edit1 and you click cancel of Edit154, the main composer should not show and the focus should be still on Edit1. If the current focused input is something else and you click cancel, the main composer should show no matter the number of drafts.

**Comment by stephanieelliott** [comment]
@rushatgabhane do you have any feedback on this proposal? https://github.com/Expensify/App/issues/15959#issuecomment-1489160890

**Comment by tienifr** [comment]
@stephanieelliott @rushatgabhane What about my proposal above? Thanks https://github.com/Expensify/App/issues/15959#issuecomment-1480356044

**Comment by MelvinBot** [comment]
@rushatgabhane @thienlnam @stephanieelliott this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by ShogunFire** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/15959#issuecomment-1489160890)

**Comment by thienlnam** [comment]
bump @rushatgabhane Could you please provide an update here?

**Comment by rushatgabhane** [comment]
Asked @s77rt if they can help https://expensify.slack.com/archives/C02NK2DQWUX/p1680808621691939

**Comment by rushatgabhane** [comment]
@thienlnam could you please assign @s77rt to review, thanks!

**Comment by MelvinBot** [comment]
@thienlnam, @stephanieelliott Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@thienlnam, @stephanieelliott Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
📣 @s77rt You have been assigned to this job by @thienlnam!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by MelvinBot** [comment]
Current assignee @s77rt is eligible for the Internal assigner, not assigning anyone new.

**Comment by s77rt** [comment]
Not overdue, still holding for Slack discussion.

**Comment by s77rt** [comment]
Not overdue, just bumped the slack thread for more :eyes:.

**Comment by s77rt** [comment]
Still waiting for clear outcome from the slack discussion

**Comment by MelvinBot** [comment]
@s77rt, @thienlnam, @stephanieelliott Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Hi @stephanieelliott :wave: There is some recent comments on that discussion. So I think we should still hold this for a while?

**Comment by thienlnam** [comment]
From the discussion, we're going to look to having RN updated to provide us the things we need for focus

**Comment by thienlnam** [comment]
@s77rt Want to lead up that effort upstream?

**Comment by s77rt** [comment]
@thienlnam Yes, will work on that asap.

**Comment by s77rt** [comment]
I have raised a proposal here https://github.com/react-native-community/discussions-and-proposals/issues/640 Hopefully we will get some answers / guidance.

**Comment by s77rt** [comment]
No activity in the linked proposal yet

**Comment by s77rt** [comment]
Same ^

**Comment by stephanieelliott** [comment]
Moving this to `Weekly` while we work through the RN updates, since this issue is blocked on that

**Comment by s77rt** [comment]
Not overdue. Still on hold

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
@stephanieelliott I'm not totally sure how community proposals work. It's likely to take a very long time. Do you think we should follow a different approach here?

**Comment by stephanieelliott** [comment]
No, not necessarily! This seems like it may be a situation where we need to just wait. Just wanted to make sure there was not some extra context I was missing 😊

**Comment by stephanieelliott** [comment]
Still on hold, moving this to monthly since we'll likely not have frequent updates.

**Comment by s77rt** [comment]
Still on hold

**Comment by s77rt** [comment]
Still on hold but I think we may be fixing this in https://github.com/Expensify/App/issues/23908

**Comment by stephanieelliott** [comment]
Oh great, sounds like https://github.com/Expensify/App/issues/23908 may fix this! Will keep an eye there.

**Comment by stephanieelliott** [comment]
https://github.com/Expensify/App/issues/23908 has been merged, just re-tested this and seems to be fixed now
